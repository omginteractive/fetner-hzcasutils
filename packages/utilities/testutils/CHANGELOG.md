# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 1.2.3 (2020-04-20)


### @hzdg/commitizen-config

* fix defaults when no config exists at all ([c31252e](https://github.com/hzdg/hz-core/commit/c31252e))


## 1.2.2 (2020-04-18)


* add netlify-preinstall script ([8b87c66](https://github.com/hzdg/hz-core/commit/8b87c66))


## [1.2.1](https://github.com/hzdg/hz-core/compare/testutils@1.2.0...testutils@1.2.1) (2020-04-15)


* namespace under @hzdg ([b1b92f3](https://github.com/hzdg/hz-core/commit/b1b92f3))


# [1.2.0](https://github.com/hzdg/hz-core/compare/testutils@1.1.0...testutils@1.2.0) (2020-02-24)


### testutils

* add KeyboardSequence modifier methods ([0b7b50e](https://github.com/hzdg/hz-core/commit/0b7b50e))
* fix EventSequence run queue continuation ([c217066](https://github.com/hzdg/hz-core/commit/c217066))
* add return types ([f487679](https://github.com/hzdg/hz-core/commit/f487679))


# [1.1.0](https://github.com/hzdg/hz-core/compare/testutils@1.0.1...testutils@1.1.0) (2020-02-10)


### testutils

* add repeat() to {Mouse,Touch,Wheel}Sequence ([b0ebe9a](https://github.com/hzdg/hz-core/commit/b0ebe9a))
* fix getComputedValue overload types ([bfc6031](https://github.com/hzdg/hz-core/commit/bfc6031))


## [1.0.1](https://github.com/hzdg/hz-core/compare/testutils@1.0.0...testutils@1.0.1) (2019-11-08)


### testutils

* move MockMutationObserver to testutils ([a87950f](https://github.com/hzdg/hz-core/commit/a87950f))
* fix type errors ([0743e7a](https://github.com/hzdg/hz-core/commit/0743e7a))


## [1.0.0](https://github.com/hzdg/hz-core/compare/testutils@0.0.2...testutils@1.0.0) (2019-05-20)


### testutils

* fix EventSequence API and types ([2317f2f](https://github.com/hzdg/hz-core/commit/2317f2f))


## 0.0.2 (2019-05-03)


### testutils

* move testutils to their own package ([22ba155](https://github.com/hzdg/hz-core/commit/22ba155))
