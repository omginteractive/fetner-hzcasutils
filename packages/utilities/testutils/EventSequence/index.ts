/* eslint-env jest */
export {default, default as EventSequence} from './EventSequence';
export {default as MouseSequence} from './MouseSequence';
export {default as TouchSequence} from './TouchSequence';
export {default as KeyboardSequence} from './KeyboardSequence';
export {default as WheelSequence} from './WheelSequence';
export {default as GestureHistory} from './GestureHistory';
