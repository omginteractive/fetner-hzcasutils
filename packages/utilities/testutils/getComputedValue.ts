/**
 * `getComputedValue` will return a getter for a value
 * from an Element's computed styles.
 */
function getComputedValue<K extends keyof CSSStyleDeclaration>(
  /** The name of the computed style to get, i.e., 'width'. */
  value: K,
): (this: Element) => CSSStyleDeclaration[K];
/**
 * `getComputedValue` will return a getter for a value
 * from an Element's computed styles cast as `Number`.
 */
function getComputedValue<K extends keyof CSSStyleDeclaration>(
  /** The name of the computed style to get, i.e., 'width'. */
  value: K,
  as: typeof Number,
): (this: Element) => number;
/**
 * `getComputedValue` will return a getter for a value
 * from an Element's computed styles cast as `Boolean`.
 */
function getComputedValue<K extends keyof CSSStyleDeclaration>(
  /** The name of the computed style to get, i.e., 'width'. */
  value: K,
  as: typeof Boolean,
): (this: Element) => boolean;
function getComputedValue<
  K extends keyof CSSStyleDeclaration,
  T extends typeof Number | typeof Boolean
>(value: K, as?: T) {
  return function(this: Element) {
    const computed = getComputedStyle(this)[value];
    switch (as) {
      case Number:
        return parseFloat(computed);
      case Boolean:
        return Boolean(computed);
      default:
        return computed;
    }
  };
}

export default getComputedValue;
