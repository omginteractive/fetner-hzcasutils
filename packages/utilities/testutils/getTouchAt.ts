import hasProperty from './hasProperty';

export default function getTouchAt(
  from: TouchList,
  index: number,
): Touch | null {
  return hasProperty(from, 'item') ? from.item(index) : from[index];
}
