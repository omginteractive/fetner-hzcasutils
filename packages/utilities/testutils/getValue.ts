function getValue<T, K extends keyof T>(obj: T, key: K): T[K];
function getValue<T, K extends keyof T, V>(
  obj: T,
  key: K,
  defaultValue: V,
): Exclude<T[K], undefined> | V;
function getValue<T, K extends keyof T, V>(
  obj: T,
  key: K,
  defaultValue?: V,
): T[K] | V | undefined {
  const value = obj[key];
  return value === undefined ? defaultValue : value;
}

export default getValue;
