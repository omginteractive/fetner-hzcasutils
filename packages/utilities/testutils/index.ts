export {default as sleep} from './sleep';
export {default as hasProperty} from './hasProperty';
export {default as getFlag} from './getFlag';
export {default as getValue} from './getValue';
export {default as getTouchAt} from './getTouchAt';
export {default as getComputedValue} from './getComputedValue';
