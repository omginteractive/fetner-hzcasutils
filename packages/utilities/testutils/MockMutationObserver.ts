/* eslint-env jest, browser */
import getComputedValue from './getComputedValue';

// We have to mock these aspects of HTMLElement because
// the underlying ResizeObserver polyfill uses them,
// but they're all `0` by default, resulting in no resize broadcasts!
const elementSpies: jest.SpyInstance[] = [];
beforeEach(() => {
  /**
   * Globally mock MutationObserver so it can be advanced using jest's fake timers.
   * We mock this because it is used by the ResizeObserver polyfill
   * to schedule microtasks.
   * See any of these to find out how deep the rabbit hole goes.
   * https://github.com/facebook/jest/issues/2157
   * https://github.com/facebook/jest/issues/7151
   * https://github.com/sinonjs/lolex/issues/114
   * https://gist.github.com/quasicomputational/1651489425f7f3b3a918ab857021d68b
   */
  // @ts-ignore
  global.MutationObserver = class MutationObserver {
    // @ts-ignore
    static original = global.MutationObserver;
    callback: MutationCallback;
    constructor(callback: MutationCallback) {
      this.callback = callback;
    }
    observe(el: Node): void {
      Object.defineProperty(el, 'textContent', {
        set: () => {
          // @ts-ignore
          setTimeout(() => this.callback([], this), 0);
        },
        configurable: true,
      });
    }
    disconnect(): void {
      this.callback = () => {};
    }
  };
  elementSpies.push(
    jest
      .spyOn(HTMLElement.prototype, 'offsetWidth', 'get')
      .mockImplementation(getComputedValue('width', Number)),
    jest
      .spyOn(HTMLElement.prototype, 'clientWidth', 'get')
      .mockImplementation(getComputedValue('width', Number)),
    jest
      .spyOn(HTMLElement.prototype, 'clientHeight', 'get')
      .mockImplementation(getComputedValue('height', Number)),
    jest
      .spyOn(HTMLElement.prototype, 'offsetHeight', 'get')
      .mockImplementation(getComputedValue('height', Number)),
  );
});

afterEach(() => {
  // @ts-ignore
  global.MutationObserver = global.MutationObserver.original;
  while (elementSpies.length) {
    const spy = elementSpies.pop();
    if (spy) spy.mockRestore();
  }
});
