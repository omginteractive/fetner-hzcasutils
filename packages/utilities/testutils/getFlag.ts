import getValue from './getValue';

export default function getFlag<T, V>(
  obj: T,
  key: keyof T,
  defaultValue?: V,
): boolean {
  return Boolean(getValue(obj, key, defaultValue));
}
