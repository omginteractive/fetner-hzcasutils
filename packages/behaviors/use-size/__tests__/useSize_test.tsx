/** @jsdom-global Object */
/* eslint-env jest, browser */
import React from 'react';
import {render, act} from '@testing-library/react';
import useSize from '../src';
import 'testutils/MockMutationObserver';

beforeEach(() => {
  jest.useFakeTimers();
});

function runAllTimers(): void {
  act(() => {
    jest.runAllTimers();
  });
}

test('useSize gets the initial size', () => {
  const SizeUser = (): JSX.Element => {
    const [size, ref] = useSize<HTMLDivElement>();
    return (
      <div ref={ref} style={{width: 120, height: 120}} data-testid="size">
        {JSON.stringify(size, null, 2)}
      </div>
    );
  };
  const {getByTestId} = render(<SizeUser />);
  runAllTimers();
  expect(getByTestId('size')).toMatchSnapshot();
});

test('useSize passes the initial size to a handler', () => {
  const handler = jest.fn();
  const SizeUser = (): JSX.Element => {
    const ref = useSize<HTMLDivElement>(handler);
    return <div ref={ref} style={{width: 120, height: 120}} />;
  };
  render(<SizeUser />);
  runAllTimers();
  expect(handler.mock.calls).toMatchSnapshot();
});

test('useSize uses an existing ref', () => {
  const SizeUser = (): JSX.Element => {
    const [node, setNode] = React.useState<HTMLDivElement | null>(null);
    const ref = React.useRef(node);
    ref.current = node;
    const size = useSize<HTMLDivElement>(ref);
    return (
      <div
        ref={node => setNode(node)}
        style={{width: 120, height: 120}}
        data-testid="size"
      >
        {JSON.stringify(size, null, 2)}
      </div>
    );
  };
  const {getByTestId} = render(<SizeUser />);
  runAllTimers();
  expect(getByTestId('size')).toMatchSnapshot();
});

test('useSize uses an existing ref and a handler', () => {
  const handler = jest.fn();
  const SizeUser = (): JSX.Element => {
    const [node, setNode] = React.useState<HTMLDivElement | null>(null);
    const ref = React.useRef(node);
    ref.current = node;
    useSize<HTMLDivElement>(ref, handler);
    return (
      <div ref={node => setNode(node)} style={{width: 120, height: 120}} />
    );
  };
  render(<SizeUser />);
  runAllTimers();
  expect(handler.mock.calls).toMatchSnapshot();
});

test('useSize resubscribes to changed refs', () => {
  const SizeUser = ({index}: {index: number}): JSX.Element => {
    const [size, ref] = useSize<HTMLDivElement>();
    return (
      <div data-testid="size">
        <div ref={index ? null : ref} style={{width: 1, height: 1}}>
          {index ? 'see below' : JSON.stringify(size, null, 2)}
        </div>
        <div ref={index ? ref : null} style={{width: 2, height: 2}}>
          {index ? JSON.stringify(size, null, 2) : 'see above'}
        </div>
      </div>
    );
  };
  const {getByTestId, rerender} = render(<SizeUser index={0} />);
  runAllTimers();
  expect(getByTestId('size')).toMatchSnapshot();
  rerender(<SizeUser index={1} />);
  runAllTimers();
  expect(getByTestId('size')).toMatchSnapshot();
});

test('useSize works on iframed elements', () => {
  const iframe = document.createElement('iframe');
  document.body.appendChild(iframe);
  const node = (iframe.contentDocument as Document).createElement('div');
  const SizeUser = (): JSX.Element => {
    const [size, ref] = useSize<HTMLDivElement>();
    return (
      <div ref={ref} style={{width: 120, height: 120}} data-testid="size">
        {JSON.stringify(size, null, 2)}
      </div>
    );
  };
  const {getByTestId} = render(<SizeUser />, {container: node});
  runAllTimers();
  expect(getByTestId('size')).toMatchSnapshot();
});
