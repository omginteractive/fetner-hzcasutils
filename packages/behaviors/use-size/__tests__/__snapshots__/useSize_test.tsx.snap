// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSize gets the initial size 1`] = `
<div
  data-testid="size"
  style="width: 120px; height: 120px;"
>
  {
  "x": 0,
  "y": 0,
  "top": 0,
  "right": 120,
  "bottom": 120,
  "left": 0,
  "width": 120,
  "height": 120,
  "vw": 0.12,
  "vh": 0.16,
  "borderBoxSize": {
    "inlineSize": 120,
    "blockSize": 120
  },
  "contentBoxSize": {
    "inlineSize": 120,
    "blockSize": 120
  }
}
</div>
`;

exports[`useSize passes the initial size to a handler 1`] = `
Array [
  Array [
    Object {
      "borderBoxSize": Object {
        "blockSize": 120,
        "inlineSize": 120,
      },
      "bottom": 120,
      "contentBoxSize": Object {
        "blockSize": 120,
        "inlineSize": 120,
      },
      "height": 120,
      "left": 0,
      "right": 120,
      "top": 0,
      "vh": 0.16,
      "vw": 0.12,
      "width": 120,
      "x": 0,
      "y": 0,
    },
  ],
]
`;

exports[`useSize resubscribes to changed refs 1`] = `
<div
  data-testid="size"
>
  <div
    style="width: 1px; height: 1px;"
  >
    {
  "x": 0,
  "y": 0,
  "top": 0,
  "right": 1,
  "bottom": 1,
  "left": 0,
  "width": 1,
  "height": 1,
  "vw": 0,
  "vh": 0,
  "borderBoxSize": {
    "inlineSize": 1,
    "blockSize": 1
  },
  "contentBoxSize": {
    "inlineSize": 1,
    "blockSize": 1
  }
}
  </div>
  <div
    style="width: 2px; height: 2px;"
  >
    see above
  </div>
</div>
`;

exports[`useSize resubscribes to changed refs 2`] = `
<div
  data-testid="size"
>
  <div
    style="width: 1px; height: 1px;"
  >
    see below
  </div>
  <div
    style="width: 2px; height: 2px;"
  >
    {
  "x": 0,
  "y": 0,
  "top": 0,
  "right": 2,
  "bottom": 2,
  "left": 0,
  "width": 2,
  "height": 2,
  "vw": 0,
  "vh": 0,
  "borderBoxSize": {
    "inlineSize": 2,
    "blockSize": 2
  },
  "contentBoxSize": {
    "inlineSize": 2,
    "blockSize": 2
  }
}
  </div>
</div>
`;

exports[`useSize uses an existing ref 1`] = `
<div
  data-testid="size"
  style="width: 120px; height: 120px;"
>
  {
  "x": 0,
  "y": 0,
  "top": 0,
  "right": 120,
  "bottom": 120,
  "left": 0,
  "width": 120,
  "height": 120,
  "vw": 0.12,
  "vh": 0.16,
  "borderBoxSize": {
    "inlineSize": 120,
    "blockSize": 120
  },
  "contentBoxSize": {
    "inlineSize": 120,
    "blockSize": 120
  }
}
</div>
`;

exports[`useSize uses an existing ref and a handler 1`] = `
Array [
  Array [
    Object {
      "borderBoxSize": Object {
        "blockSize": 120,
        "inlineSize": 120,
      },
      "bottom": 120,
      "contentBoxSize": Object {
        "blockSize": 120,
        "inlineSize": 120,
      },
      "height": 120,
      "left": 0,
      "right": 120,
      "top": 0,
      "vh": 0.16,
      "vw": 0.12,
      "width": 120,
      "x": 0,
      "y": 0,
    },
  ],
]
`;

exports[`useSize works on iframed elements 1`] = `
<div
  data-testid="size"
  style="width: 120px; height: 120px;"
>
  {
  "x": 0,
  "y": 0,
  "top": 0,
  "right": 0,
  "bottom": 0,
  "left": 0,
  "width": 0,
  "height": 0,
  "vw": 0,
  "vh": 0,
  "borderBoxSize": {
    "blockSize": 0,
    "inlineSize": 0
  },
  "contentBoxSize": {
    "blockSize": 0,
    "inlineSize": 0
  }
}
</div>
`;
